docs/VirtualMachine.md

```diff
@@ -0,0 +1,160 @@
+# Virtual Machine Design
+
+This document details the workings and design of the virtual machine that executes a Flux script.
+
+## Execution Stack
+
+A Flux script goes through several transformations from text to execution.
+
+1. Flux source code is created via any means.
+2. The parser produces an abstract syntax tree (AST) from the source code.
+3. The semantic analyzer produces a semantic graph from the AST.
+4. The compiler accepts the semantic graph and produces an intermediate representation (IR) to be executed.
+5. The flux virtual machine executes the IR.
+
+NOTE: Should we call it bytecode or IR?
+Should we call it a VM or an interpreter?
+
+## Execution model
+
+Flux has a specific execution model.
+The model depends on the definitions of various data types and operations on those types.
+The Flux spec defines some of those types.
+This document reviews those types as well as defines a few new types that are specific to the implementation.
+
+### Data types and operations
+
+#### Table
+
+A Flux table consists of rows and columns.
+Columns have a label and a type.
+The table has zero or more rows, where each row defines a value for each column or a null is used to indicate a missing value.
+
+A table additionaly has a group key, which is a list of columns and their values.
+All rows within a table have the same value for all columns that are part of the group key.
+
+#### Stream
+
+A stream is an unbounded set of tables.
+Each table must have a different group key.
+
+#### Input
+
+An input is a source of a stream.
+
+Inputs are reprented as functions that produce a stream.
+
+#### Output
+
+An output consumes a stream and send the data to an external system.
+
+Output are reprented as functions that consume a stream and produce a stream.
+Outputs are transparent proxies, passing their input data downstream unmodified.
+As a result outputs themselves have no effect on the Flux program, in other words they only have side effects.
+
+#### Transformation
+
+A transformation is not a data type but rather an operation defined on a stream.
+A transformation consumes a stream of tables and produces a new stream of tables.
+
+Transformations may make use of a data cache to allow for multiple edits to a table before materializing the table into the new stream.
+
+Transformations are reprented as functions that consume a stream and produce a stream.
+
+#### Data cache
+
+A data cache is a set of tables that have yet to be materialized.
```

Comment:
I find this statement confusing.  I think of "materialization" as happening anytime a table is stored with all its data accessible.  As a counterexample, in a sequence of operations that can be pipelined such that they operate on a single row at a time, I would say the table would *not* be materialized between such operations.

By "materialized" do you mean made available to downstream transformations?